geps/gep-1651.md

```diff
@@ -0,0 +1,154 @@
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Allow users to configure a GatewayAddress so that it is only accessible within
+a specific scope (ie. public/cluster local)
+
+## Goals
+
+- Cluster local gateways can be created in a declarative way `ClusterLocal` and
+  are only accessible within a single cluster
+- Define a `Public` scope for addresses that are accessible to the internet
+- Implementations can define & support their own scopes
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an 
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "internal" 
+(service.namespace.svc) and "external" (service.namespace.example.com). 
+Gateways using the same GatewayClass on the cluster, but ensure that the 
+"internal" service is only reachable within the cluster. This would greatly 
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of 
+layer 7 load balancing for our internal services.
+
+## API
+
+We propose extending the `GatewayAddress` type to allow end-users to specify
+the accessibility of the requested address using a property named `Scope`. 
+
+`GatewayAddress.Value` is now an `optional` field. This allows users to signal
+their intent for the implementation to dynamically choose an address based on
+the attributes `GatewayAddress.Scope` and `GatewayAddress.Type`.
+
+### Predefined Scopes
+
+Implementations MAY implement the following supplied scopes and MUST abide by 
+their defined accessibility
+
+Scope | Accessibility
+-|-
+`Public`|The address is accessible over the public internet
+`ClusterLocal`|The address is accessible only within the local cluster
+
+### Vendor prefixed scopes
+
+Implementations can define custom scopes by specifying a vendor prefix followed
+by a slash `/` and a custom scope name ie. `com.ingress.dev/my-scope`.
+
+### Default scope
+The default value of `scope` is implementation specific. It is RECOMMENDED that
+the default `scope` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+
+## Examples
+
+#### 1. Request a GatewayAddress that is only accessible within a cluster
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  listeners:  
+  - protocol: HTTP
+    port: 80
+  addresses:
+  - scope: ClusterLocal
+```
+#### 2. Request a GatewayAddress with a specific scope and address
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  listeners:  
+  - protocol: HTTP
+    port: 80
+  addresses:
+  - scope: ClusterLocal
+    value: 10.0.0.8
```

Comment:
What's supposed to happen if there's already a ClusterIP Service that has claimed this IP?

Reply:
Service is denied, causing the Gateway to not be Programmed.

Example should probably use Public where its more common to configure the IP

Reply:
Yeah I agree - you could say then the Gateway `Accepted` status becomes `False` then `Ready` `False`

Reply:
What if the referenced `gatewayclass.spec.parametersRef` specifies Gateway addresses? Would these addresses take precedence for the Gateway?